config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_containerclusters.container.cnrm.cloud.google.com.yaml

```diff
@@ -1808,6 +1808,24 @@ spec:
                   current reported status reflects the most recent desired state of
                   the resource.
                 type: integer
+              observedState:
```

Comment:
So given that e.g. masterVersion is status.masterVersion, do we want to introduce status.observedState?  i.e. should it be status.observedState.masterAuth.clientCertificate vs status.masterAuth.clientCertificate?

The way I see it:
1. If we introduce status.observedState we have to deal with the status.masterVersion vs status.observedState.masterVersion inconsistency at some time in the future
2. If we don't introduce status.observedState we (maybe) have a naming conflict if there's a status field in the API called "conditions"

Personally, I'd rather deal with the hypothetical work of (2) than the known work of (1), but that's just my intuition

Reply:
Thank you for the suggestions, @justinsb !

> 1. If we introduce status.observedState we have to deal with the `status.masterVersion` vs `status.observedState.masterVersion` inconsistency at some time in the future
>
> 2. If we don't introduce `status.observedState` we (maybe) have a naming conflict if there's a status field in the API called "conditions"

IMO, if we introduce `status.observedState.masterVersion` one day, then `status.masterVersion` and `status.observedState.masterVersion` will always share the same value. I'd consider it a redundancy but not inconsistency. Or are you referring to some other scenarios?

If we need to deal with naming conflict, then it seems to be a true inconsistency: possibly we either need to rename the observed field (inconsistent with its original field name), or rename the existing status field (may impact all CRDs).

And in general I find redundancy safer to the users.

Reply:
My 2 cents. I think grouping the observed state has additional benefits in that we're grouping it. We can tell customers (and future devs) that all cloud response state can be found under .spec.observedState. More importantly I think we should insist that only cloud response state is allowed into .spec.observedState. That allows anyone looking at a value under .spec.observedState to trust its a value from the cloud provider and not say an artifact from the controller. I think it will make a docs easier and clearer. I think the fact that it essentially name spaces the response data and hence provides protection from future conflicts (either from new fields, newly cared about response fields or other non cloud response status fields which happen to have a name collision) is valuable but possibly not the most valuable.

Reply:
I agree with @cheftako, my view on a dedicated `status.observedState` is it gives our users a new mental model on how to identify status changes caused by the external cloud provider, but not KCC.

> if there's a status field in the API called "conditions"

Also I want to mention this is not a hypothetical scenario. We have already seen it in a few CRDs. See the following two examples in our servicemapping config. We had to ignore a bunch of fields because they have conflict naming with the reserved K8s status name fields.

https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/config/servicemappings/run.yaml#L27

https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/config/servicemappings/run.yaml#L147

Should customers require support for these ignored fields in the future, we have the flexibility to add them to `status.observedState`.

Reply:
Thank you for the input, @cheftako and @diviner524 !

@justinsb I'll probably merge the implementation as is for now. If you think not having `status.observedState` is important, then let's discuss offline about a revised design.